### PR TITLE
GPIO Precharge

### PR DESCRIPTION
**Pull request by SoroushF79**
Layer Precharge on currently unwritten GPIO functions

### CODE REVIEW

File: BSP/Simulator/Hardware/PreCharge.py [hunk 1]
```diff
@@ -3,10 +3,13 @@
...
 import fcntl
+import GPIO
  
-file = "BSP/Simulator/Hardware/Data/PreCharge.csv";
-
+file = "BSP/Simulator/Hardware/Data/GPIO.csv";
 
+def write(data, port=2):
+    GPIO_write(data, port)
+
```
Comment (JimothyGreene) [inline BSP/Simulator/Hardware/PreCharge.py:12]: There won't be a `GPIO.py`, so this write function should write the data to the correct line of the `GPIO.csv` itself

File: BSP/Simulator/Src/BSP_Precharge.c [hunk 2]
```diff
@@ -14,43 +14,15 @@
...
-        printf("Precharge file is not found\n");
-        return;
-    }
-
-    int fno = fileno(fp);
-    flock(fno,LOCK_EX);
-
-    int data = 0;
-    fscanf(fp,"%d", &data);
+    int data = BSP_GPIO_Read();
```
Comment (JimothyGreene) [inline BSP/Simulator/Src/BSP_Precharge.c:17]: You'll end up passing in a parameter here. It'll probably be an enum, so don't worry about it for now. Just a heads up

**Review by JimothyGreene — CHANGES_REQUESTED** [review]
Few minor changes for this prototyped version

File: Drivers/Src/Precharge.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
+/* Copyright (c) 2020 UT Longhorn Racing Solar */
+
+#include "Precharge.h"
+
+#define FILE_NAME DATA_PATH(PRECHARGE_CSV)
+#define PRECHARGE_PORT PORTC
+#define MOTOR_BIT 3
```
Comment (ErickCortez98) [inline Drivers/Src/Precharge.c:7]: These could be part of the header file

**Review by ErickCortez98 — CHANGES_REQUESTED** [review]

**Review by JimothyGreene — CHANGES_REQUESTED** [review]
The Python simulation side doesn't read from the correct port (line) in the GPIO.csv file. It should be reading from line 3 (port C). Also, we don't need a write function since we'll never change the state of the precharge boards from the simulator. I would take a look at the Contactors code that was merged into master recently. It should be nearly identical on both the Python and C sides.

**Review by JimothyGreene — APPROVED** [review]
All working!

**Review by ErickCortez98 — APPROVED** [review]

**Pull request merged**